### PR TITLE
ci(lint): remove GitHub workflow runtime deprecations, ensure lint is up to date

### PR DESCRIPTION
**Pull request by ramin**
## Description

Decided to take a look at #1000 and see what was up, seems linting DOES run and is not catching issues that are passed locally, but, while here i noticed some very out of date deprecation warnings on github action runtimes. So:

- updated `actions/checkout` to `v3` -> `v4`
- updated `actions/setup-go` to `v4` -> `v5`
- noticed the `git diff` action was stuck on node16 runtime and deprecated, so replicated behavior with paths in workflow trigger, even though it does require 2 entries (1 for push, 1 for pull_request) seemed cleaner than an unmaintained github action workflow
- finally: set `setup-go` to use `go.mod` for sourcing the go version.

If welcome, i intend to update warnings across all workflows next.

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use
  [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Review by evan-forbes — APPROVED** [review]

**Review by rootulp — DISMISSED** [review]
Thanks for bumping the outdated deps. I don't understand how this closes https://github.com/celestiaorg/celestia-core/issues/1000 because it looks like none of the lint issues identified in that issue are actually addressed in this PR.

**Review by ninabarbakadze — APPROVED** [review]

**Comment by ramin** [comment]
> Thanks for bumping the outdated deps. I don't understand how this closes #1000 because it looks like none of the lint issues identified in that issue are actually addressed in this PR.

@rootulp 1000 was the trigger to try and look, but couldn't repro those warnings/errors anymore so cleaned up the rest while in there.

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -10,32 +10,36 @@ name: Golang Linter
...
+      - uses: actions/checkout@v4
+      
+      - uses: actions/setup-go@v5
         with:
-          PATTERNS: |
-            **/**.go
-            go.mod
-            go.sum
-      - uses: golangci/golangci-lint-action@v3.7.0
+          go-version-file: 'go.mod'
```
Comment (rootulp) [inline .github/workflows/lint.yml:36]: [no change needed] 😮 clever! We can remove a few explicit Go versions from CI jobs in celestia-app with this trick.

Ref: https://github.com/actions/setup-go?tab=readme-ov-file#getting-go-version-from-the-gomod-file

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -10,32 +10,36 @@ name: Golang Linter
...
+
+      - uses: golangci/golangci-lint-action@v4.0.0
         with:
           # Required: the version of golangci-lint is required and
           # must be specified without patch version: we always use the
           # latest patch version.
           version: v1.55.2
```
Comment (rootulp) [inline .github/workflows/lint.yml:43]: [optional] we could get rid of this comment b/c it's incorrect and bump to latest golangci-lint https://github.com/golangci/golangci-lint/releases/tag/v1.56.2

but we can def do this in a follow up

**Review by rootulp — APPROVED** [review]

**Review by staheri14 — APPROVED** [review]
LGTM!

**Comment by rootulp** [comment]
Agreed. Updated the PR description.

**Pull request merged**